checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java

```diff
@@ -0,0 +1,199 @@
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.ReturnTree;
+import com.sun.source.tree.Tree;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import org.checkerframework.checker.compilermsgs.qual.CompilerMessageKey;
+import org.checkerframework.checker.mustcall.qual.MustCallAlias;
+import org.checkerframework.checker.mustcall.qual.NotOwning;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.basetype.BaseTypeVisitor;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedDeclaredType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType;
+import org.checkerframework.javacutil.ElementUtils;
+import org.checkerframework.javacutil.TreePathUtil;
+import org.checkerframework.javacutil.TreeUtils;
+
+/**
+ * The visitor for the Must Call Checker. This visitor is similar to BaseTypeVisitor, but overrides
+ * methods that don't work well with the MustCall type hierarchy because it doesn't use the top type
+ * as the default type.
+ */
+public class MustCallVisitor extends BaseTypeVisitor<MustCallAnnotatedTypeFactory> {
+
+  /**
+   * Creates a new MustCallVisitor.
+   *
+   * @param checker the type-checker associated with this visitor
+   */
+  public MustCallVisitor(BaseTypeChecker checker) {
+    super(checker);
+  }
+
+  @Override
+  public Void visitReturn(ReturnTree node, Void p) {
+    // Only check return types if ownership is being transferred.
+    if (!checker.hasOption(MustCallChecker.NO_LIGHTWEIGHT_OWNERSHIP)) {
+      MethodTree enclosingMethod = TreePathUtil.enclosingMethod(this.getCurrentPath());
+      // enclosingMethod is null if this return site is inside a lambda. TODO: handle lambdas more
+      // precisely?
+      if (enclosingMethod != null) {
+        ExecutableElement methodElt = TreeUtils.elementFromDeclaration(enclosingMethod);
+        AnnotationMirror notOwningAnno = atypeFactory.getDeclAnnotation(methodElt, NotOwning.class);
+        if (notOwningAnno != null) {
+          // Skip return type subtyping check, because not-owning pointer means Object Construction
+          // Checker won't check anyway.
+          return null;
+        }
+      }
+    }
+    return super.visitReturn(node, p);
+  }
+
+  @Override
+  public boolean isValidUse(
+      AnnotatedDeclaredType declarationType, AnnotatedDeclaredType useType, Tree tree) {
+    // MustCallAlias annotations are always permitted on type uses, because these will be validated
+    // by the Object Construction Checker's -AcheckMustCall algorithm.
+    if (!checker.hasOption(MustCallChecker.NO_RESOURCE_ALIASES)) {
+      useType.removeAnnotationByClass(MustCallAlias.class);
```

Comment:
I am concerned that this side-effects the type that was passed in.  Are you positive that this will never create a problem elsewhere?  Maybe this method should make a copy.